tests/Simplify/CaseOfTest.elm

```diff
@@ -377,6 +377,130 @@ a =
+                            { message = "Unnecessary cases"
+                            , details = [ "The value between case ... of is a known list of length 0. However, the 1st and 2nd and 3rd case matches on a list with a different length which means you can remove it." ]
+                            , under = "_ :: _ :: _"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a =
+    case () of
+
+
+
+        () ->
+            3
+"""
+                        ]
+        , test "should remove unnecessary case of filled list literal when all cases are list patterns" <|
+            \() ->
+                """module A exposing (..)
+a =
+    case [ 0, 0 ] of
+        _ :: _ :: _ ->
+            0
+        
+        [ _ ] ->
+            1
+        
+        [ _, _ ] ->
+            2
+        
+        [ _, _, _ ] ->
+            3
+
+        _ :: _ ->
+            4
+        
+        [] ->
+            5
+"""
+                    |> Review.Test.run ruleWithDefaults
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "Unnecessary cases"
+                            , details = [ "The value between case ... of is a known list of length 2. However, the 2nd and 4th and 6th case matches on a list with a different length which means you can remove it." ]
+                            , under = "[ _ ]"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a =
+    case (0, [ 0 ]) of
+        (_, _ :: _) ->
+            0
+        
+
+        (_, [ _ ]) ->
+            2
+        
+
+        (_, _) ->
+            4
+        
+"""
+                        ]
+        , test "should remove unnecessary case of filled list literal when all cases are list patterns with length >= 3" <|
+            \() ->
+                """module A exposing (..)
+a =
+    case [ 0, 0, 0, 0 ] of
+        [ _, _, _, _ ] ->
+            0
+        
+        _ :: _ :: _ :: _ ->
+            1
+        
+        [] ->
+            2
+        
+        [ _ ] ->
+            3
+        
+        [ _, _ ] ->
+            4
+        
+        [ _, _, _ ] ->
+            5
+"""
+                    |> Review.Test.run ruleWithDefaults
+                    |> Review.Test.expectErrors
+                        [ Review.Test.error
+                            { message = "Unnecessary cases"
+                            , details = [ "The value between case ... of is a known list of length 4. However, the 3rd and 4th and 5th and 6th case matches on a list with a different length which means you can remove it." ]
+                            , under = "[]"
+                            }
+                            |> Review.Test.whenFixed """module A exposing (..)
+a =
+    case ((0, (0, 0)), [ 0 ]) of
+        ((_, (_, _)), [ _ ]) ->
+            0
+        
+        ((_, (_, _)), _) ->
```

Comment:
First: Currently, we don't simplify case-ofs with wildcard patterns at all, see the PR body section.
Second: Not converting the (beginning of) list to a tuple defeats the whole purpose of the check IMO.
In my eyes, there is no difference between leaving `case multi-variant of` or `case list with some elements of` instead of matching on a structure where we know for a fact that the values exist and we don't need these ugly (potentially unnecessary) cases.

If this is only about readability, maybe the new nesting is better?